substrate/frame/distribution/src/functions.rs

```diff
@@ -0,0 +1,163 @@
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Helper functions for Distribution pallet.
+
+pub use super::*;
+impl<T: Config> Pallet<T> {
+	pub fn pot_account() -> AccountIdOf<T> {
+		// Get Pot account
+		let pot_id = T::PotId::get();
+		let pot_account: AccountIdOf<T> = pot_id.into_account_truncating();
+		pot_account
+	}
+
+	pub fn get_spend(project_account: ProjectId<T>) -> Vec<SpendIndex> {
+		let mut spends: Vec<SpendIndex> = Vec::new();
+		let value = Some(project_account);
+
+		for spend in Spends::<T>::iter() {
+			let info = spend.1;
+			if info.whitelisted_project == value {
+				spends.push(spend.0);
+			}
+		}
+
+		spends
+	}
+
+	/// Series of checks on the Pot, to ensure that we have enough funds
+	/// before executing a Spend
+	pub fn pot_check(amount: BalanceOf<T>) -> DispatchResult {
+		// Get Pot account
+		let pot_account: AccountIdOf<T> = Self::pot_account();
+
+		// Check that the Pot as enough funds for the transfer
+		let balance = T::NativeBalance::balance(&pot_account);
+		let minimum_balance = T::NativeBalance::minimum_balance();
+		let remaining_balance = balance.saturating_sub(amount);
+
+		ensure!(remaining_balance > minimum_balance, Error::<T>::InsufficientPotReserves);
+		ensure!(balance > amount, Error::<T>::InsufficientPotReserves);
+		Ok(())
+	}
+
+	/// Funds transfer from the Pot to a project account
+	pub fn spend(
+		amount: BalanceOf<T>,
+		beneficiary: AccountIdOf<T>,
+		spend_index: u32,
+	) -> DispatchResult {
+		// Get Pot account
+		let pot_account: AccountIdOf<T> = Self::pot_account();
+
+		//Operate the transfer
+		let result =
+			T::NativeBalance::transfer(&pot_account, &beneficiary, amount, Preservation::Preserve)
+				.map_err(|_| Error::<T>::TransferFailed);
+
+		Self::process_failed_spend_result(spend_index, result)?;
+
+		Ok(())
+	}
+
+	/// Helper function used to change the status of a failed Spend
+	pub fn process_failed_spend_result(
+		spend_index: u32,
+		result: Result<BalanceOf<T>, Error<T>>,
+	) -> Result<BalanceOf<T>, Error<T>> {
+		match result {
+			Ok(x) => {
+				// Change Spend status
+				Spends::<T>::mutate(spend_index, |val| {
+					let mut val0 = val.clone().unwrap();
+					val0.status = SpendState::Completed;
+					*val = Some(val0);
+				});
+				Ok(x)
+			},
+			Err(_e) => {
+				// Change Spend status
+				Spends::<T>::mutate(spend_index, |val| {
```

Comment:
I did not trace the calls but usually when you return errors, storage operations are all reverted so this is pointless
It could be somewhere you have handled the error so this code is fine, but the fact that I am not sure if this code is correct or not means this code can be improved. maybe some comment will do, or some refactor

Reply:
Actually, because we are reserving the funds to be used first, a spending action not going through/failing is unlikely. In this case I agree that managing the status of a failing spend is dead code. I will try to think about a test case leading to this scenario, but If I don't find any, I will remove the `SpendState::Failed` status an the related code.